### PR TITLE
Harden emote parsing

### LINKED ISSUE #3487: `/fakemsg` command crashes the application
### Checklist

- [X] I'm reporting a problem with Chatterino
- [X] I've verified that I'm running **the most** recent nightly build or stable release
- [X] I've looked for my problem on the [wiki](https://wiki.chatterino.com/Help/)
- [X] I've searched the [issues and pull requests](https://github.com/Chatterino/chatterino2/issues?q=) for similar looking reports

### Describe your issue

If you use /fakemsg and you somehow mess up, the debug version of Chatterino crashes.

 "Debug Assertion Failed!"

Expression: vector subscript out of range.

And pressing "Retry" or "Ignore" closes the application, no matter what.

### Screenshots

_No response_

### OS and Chatterino Version

built from master branch

### PR DESCRIPTION
**Pull request by pajlada**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description

Messages I tested with

#### START OUT OF RANGE
`/fakemsg @emotes=84608:9-10 :test!test@test.tmi.twitch.tv PRIVMSG #pajlada :foo bar`

#### SINGLE CHARACTER EMOTE
`/fakemsg @emotes=84608:0-0 :test!test@test.tmi.twitch.tv PRIVMSG #pajlada :foo bar`

#### TWO CHARACTER EMOTE
`/fakemsg @emotes=84608:0-1 :test!test@test.tmi.twitch.tv PRIVMSG #pajlada :foo bar`

#### END OUT OF RANGE
`/fakemsg @emotes=84608:0-15 :test!test@test.tmi.twitch.tv PRIVMSG #pajlada :foo bar`

#### RANGE BAD
`/fakemsg @emotes=84608:15-2 :test!test@test.tmi.twitch.tv PRIVMSG #pajlada :foo bar`

#### MESSAGE FROM KARL-POLICE
`/fakemsg @badge-info=;badges=glhf-pledge/1;client-nonce=5d2627b0cbe56fa05faf5420def4807d;color=#1E90FF;display-name=oldcoeur;emote-only=1;emotes=84608:0-7;first-msg=1;flags=;id=7412fea4-8683-4cc9-a506-4228127a5c2d;mod=0;room-id=62300805;subscriber=0;tmi-sent-ts=1623429859222;turbo=0;user-id=139147886;user-type= :oldcoeur!oldcoeur@oldcoeur.tmi.twitch.tv PRIVMSG #test :test`

To test with these messages, join #test and #pajlada

Fixes #3487

Not adding a changelog entry as it doesn't fix an actual user-discoverable bug

### CODE REVIEW

**Review by Mm2PL — CHANGES_REQUESTED** [review]
I haven't tested yet in my c2. Would be nice if errors/malformed emotes had some kind of log and didn't fail silently. Also tests make sense for this.

**Comment by pajlada** [comment]
I haven't added a test for the `            // Emote coords are out of range from the modified character positions` error case because I can't figure out a fake message for it

**Comment by pajlada** [comment]
Issue with tests failing has to do with the `MockApplication` class symbols being merged/broken somehow - the fix is to put it in an anonymous namespace, making sure no other build object has access to it.

Will push a fix when another PR has been merged into the main branch

**Comment by pajlada** [comment]
Tests should work now :-)

File: src/providers/twitch/TwitchMessageBuilder.cpp [hunk 1]
```diff
@@ -107,6 +107,75 @@ namespace {
...
+        auto parameters = emote.split(':');
+
+        if (parameters.length() < 2)
+        {
+            return;
+        }
+
+        auto id = EmoteId{parameters.at(0)};
+
+        auto occurences = parameters.at(1).split(',');
```
Comment (Felanbird) [inline src/providers/twitch/TwitchMessageBuilder.cpp:131]: Can the other `occurence` spelling mistakes be corrected or is it already too late?

File: src/providers/twitch/TwitchMessageBuilder.cpp [hunk 2]
```diff
@@ -107,6 +107,75 @@ namespace {
...
+            TwitchEmoteOccurrence emoteOccurence{
+                start, end,
+                app->getEmotes()->getTwitchEmotes()->getOrCreateEmote(id, name),
+                name};
+            if (emoteOccurence.ptr == nullptr)
+            {
+                qCDebug(chatterinoTwitch)
+                    << "nullptr" << emoteOccurence.name.string;
+            }
+            vec.push_back(std::move(emoteOccurence));
```
Comment (Felanbird) [inline src/providers/twitch/TwitchMessageBuilder.cpp:175]: here too if possible🙂

**Pull request merged**